src/main/kotlin/oop/Observer/subscription/ObserverWithSubscription.kt

```diff
@@ -0,0 +1,79 @@
+package oop.Observer.subscription
+
+class PublishSubject<A>(private val observers: MutableList<Observer<A>> = mutableListOf()) {
+
+  fun register(observer: Observer<A>): Observer<A> {
+    observers.add(observer)
+    return observer
+  }
+
+  fun notify(value: A) {
+    observers.forEach { it.update(value) }
+  }
+
+}
+
+open class Observer<A>(private val value: A) {
+
+  private var subscription: Subscription<A>? = null
```

Comment:
But the register method is in the publisher, you only need to add another method to remove subscriber